File: geekbang/algorithm-models/3.1-排列递归模板/permute.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        used = vector<bool>(nums.size(), false);
        find(nums, 0);
        return ans;
    }


    void find(vector<int>& nums, int count) {
        if (count == nums.size()) {
            ans.push_back(s);
            return;
        }
        for (int i = 0; i < nums.size(); i++)
            if (!used[i]) {
                used[i] = true;
                s.push_back(nums[i]);
                find(nums, count + 1);
                s.pop_back();
                used[i] = false;
            }
    }


private:
    vector<vector<int>> ans;
    vector<int> s;
    vector<bool> used;
};

File: geekbang/algorithm-models/3.1-排列递归模板/permute.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        n = nums.length;
        num = new int[n];
        for (int i = 0; i < n; i++) num[i] = nums[i];
        used = new boolean[n];
        per = new ArrayList<Integer>();
        ans = new ArrayList<List<Integer>>();
        dfs(0);
        return ans;
    }


    private void dfs(int depth) {
        if (depth == n) {
            ans.add(new ArrayList<Integer>(per));
            return;
        }
        for (int i = 0; i < n; i++) {
            if (used[i]) continue;
            used[i] = true;
            per.add(num[i]);
            dfs(depth + 1);
            per.remove(per.size() - 1);
            used[i] = false;
        }
    }


    private int n;
    private int[] num;
    private boolean[] used;
    private List<Integer> per;
    private List<List<Integer>> ans;
}
